contracts/Witch.sol

```diff
@@ -13,198 +13,358 @@ import "@yield-protocol/utils-v2/contracts/math/WDivUp.sol";
-
-        Auction memory auction_ = auctions[vaultId];
-        Ilk memory ilk_ = ilks[vault_.ilkId];
-        Limits memory limits_ = limits[vault_.ilkId];
-
-        uint256 art = cauldron.debtFromBase(vault_.seriesId, base);
-        {
-            uint256 elapsed = uint32(block.timestamp) - auction_.start;                      // Auctions will malfunction on the 7th of February 2106, at 06:28:16 GMT, we should replace this contract before then.
-            uint256 price = inkPrice(balances_, ilk_.initialOffer, ilk_.duration, elapsed);
-            ink = uint256(art).wmul(price);                                                    // Calculate collateral to sell. Using divdrup stops rounding from leaving 1 stray wei in vaults.
-            require (ink >= min, "Not enough bought");
-            require (art == balances_.art || balances_.ink - ink >= limits_.dust * (10 ** limits_.dec), "Leaves dust");
-            limits[vault_.ilkId].sum -= ink.u128();
-        }
+    /// @dev Pay `base` of the debt in a vault in liquidation, getting at least `minInkOut` collateral.
+    /// Use `payAll` to pay all the debt, using `paySome` for amounts close to the whole vault might revert.
+    /// @param vaultId Id of vault to buy
+    /// @param to Receiver of the collateral bought
+    /// @param maxBaseIn Maximum amount of base that the liquidator will pay
+    /// @param minInkOut Minimum amount of collateral that must be received
+    /// @return baseIn Amount of underlying taken
+    /// @return inkOut Amount of vault collateral sold
+    function payBase(
+        bytes12 vaultId,
+        address to,
+        uint128 maxBaseIn,
+        uint128 minInkOut
+    ) external returns (uint256 baseIn, uint256 inkOut) {
+        Auction storage auction_ = auctions[vaultId];
+        require(
+            auction_.start > 0,
+            "Vault not under auction"
+        );
 
-        cauldron.slurp(vaultId, ink.u128(), art.u128());                                            // Remove debt and collateral from the vault
-        settle(msg.sender, vault_.ilkId, series_.baseId, ink.u128(), base);                   // Move the assets
-        if (balances_.art - art == 0) {                                                             // If there is no debt left, return the vault with the collateral to the owner
-            cauldron.give(vaultId, auction_.owner);
-            delete auctions[vaultId];
+        DataTypes.Vault memory vault = cauldron.vaults(vaultId);
+        DataTypes.Series memory series = cauldron.series(vault.seriesId);
+
+        // Find out how much debt is being repaid
+        uint128 artIn = uint128(cauldron.debtFromBase(vault.seriesId, maxBaseIn));
+
+        // If offering too much base, take only the necessary.
+        artIn = artIn > auction_.art ? auction_.art : artIn;
+        baseIn = cauldron.debtToBase(vault.seriesId, auction_.art);
```

Comment:
shouldn't `artIn` be used in the second param? The guy may be paying for half of the `auction.art`

Reply:
True, fixed. Thanks.